torch/csrc/distributed/rpc/rpc_with_autograd.cpp

```diff
@@ -0,0 +1,161 @@
+#include <torch/csrc/byte_order.h>
+#include <torch/csrc/distributed/rpc/utils.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+constexpr int kAutogradMessageSize = 17;
+
+AutogradMetadata::AutogradMetadata(
+    int64_t autogradContextId_,
+    int64_t autogradMessageId_)
+    : autogradContextId(autogradContextId_),
+      autogradMessageId(autogradMessageId_) {}
+
+RpcWithAutograd::RpcWithAutograd(
+    MessageType messageType,
+    const AutogradMetadata& autogradMetadata,
+    std::unique_ptr<RpcCommandBase> wrappedRpc)
+    : messageType_(messageType),
+      autogradMetadata_(autogradMetadata),
+      wrappedRpc_(std::move(wrappedRpc)) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cannot be null!");
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_REQ ||
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_RESP);
+  wrappedMessage_ = wrappedRpc_->toMessage();
+  tensors_ = wrappedMessage_.tensors();
+  wrappedMessageType_ = wrappedMessage_.type();
+}
+
+RpcWithAutograd::RpcWithAutograd(
+    MessageType messageType,
+    const AutogradMetadata& autogradMetadata,
+    std::unique_ptr<RpcCommandBase> wrappedRpc,
+    MessageType wrappedMessageType,
+    std::vector<torch::Tensor> tensors)
+    : messageType_(messageType),
+      autogradMetadata_(autogradMetadata),
+      wrappedRpc_(std::move(wrappedRpc)),
+      wrappedMessageType_(wrappedMessageType),
+      tensors_(std::move(tensors)) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cannot be null!");
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_REQ ||
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_RESP);
+}
+
+Message RpcWithAutograd::toMessage() {
+  auto payload = wrappedMessage_.movePayload();
+  TORCH_INTERNAL_ASSERT(!payload.empty());
+
+  // We append the message type (1 byte), autograd context id(8 bytes) and
+  // autograd message id(8 bytes) to the original message in network byte order
+  // (big endian).
+  size_t writableIndex = payload.size();
+
+  // Need 17 additional bytes.
+  payload.resize(payload.size() + kAutogradMessageSize);
+
+  // Add message type.
+  payload[writableIndex++] = wrappedMessage_.type();
+
+  // Add autograd ids.
+  THP_encodeInt64Buffer(
+      reinterpret_cast<uint8_t*>(payload.data()) + writableIndex,
+      &autogradMetadata_.autogradContextId,
+      THPByteOrder::THP_BIG_ENDIAN,
+      1);
+  writableIndex += sizeof(int64_t);
+  THP_encodeInt64Buffer(
+      reinterpret_cast<uint8_t*>(payload.data()) + writableIndex,
+      &autogradMetadata_.autogradMessageId,
+      THPByteOrder::THP_BIG_ENDIAN,
+      1);
+
+  return Message(
+      std::move(payload),
+      std::move(wrappedMessage_.tensors()),
+      messageType_,
+      wrappedMessage_.id());
+}
+
+std::unique_ptr<RpcWithAutograd> RpcWithAutograd::fromMessage(
+    const Message& message) {
+  TORCH_INTERNAL_ASSERT(
```

Comment:
This shouldn't really be an internal assert, since the input in this case comes from an external source, right? I.e., if I feed you garbage data I can trigger this with an assert. But that's not a bug in PyTorch.

Reply:
PyTorch end users cannot trigger this assert by feeding garbage data. The message.type() is only populated by our internal RPC implementation.